server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -426,14 +466,38 @@ public static RepositoryData snapshotsFromXContent(final XContentParser parser,
                         assert indexId != null;
                         indexSnapshots.put(indexId, snapshotIds);
                     }
+                } else if (SHARDS.equals(field)) {
+                    while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                        while (parser.nextToken() == XContentParser.Token.FIELD_NAME) {
+                            final String indexId = parser.currentName();
+                            if (parser.nextToken() != XContentParser.Token.START_ARRAY) {
+                                throw new ElasticsearchParseException("start array expected [shards]");
+                            }
+                            final Collection<String> generations = new ArrayList<>();
+                            while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                                generations.add(parser.textOrNull());
+                            }
+                            shardStatus.put(indexId, generations.toArray(Strings.EMPTY_ARRAY));
+                        }
+                    }
                 } else {
                     throw new ElasticsearchParseException("unknown field name  [" + field + "]");
                 }
             }
         } else {
             throw new ElasticsearchParseException("start object expected");
         }
-        return new RepositoryData(genId, snapshots, snapshotStates, indexSnapshots);
+        final ShardGenerations.Builder builder = ShardGenerations.builder();
+        final Map<String, IndexId> indexLookup =
+            indexSnapshots.keySet().stream().collect(Collectors.toMap(IndexId::getId, Function.identity()));
+        shardStatus.forEach((indexId, gens) -> {
+            final IndexId idx = indexLookup.get(indexId);
+            assert idx != null : "shard generations contained index [" + indexId + "] that isn't part of any snapshot";
+            for (int i = 0; i < gens.length; i++) {
+                builder.add(idx, i, gens[i]);
```

Comment:
Yes in that case the shard-level meta is updated for those shards that did successfully snapshot so that we get the references to all the new data blobs and they can be used in the next snapshot.
This is basically the same thing we have now, except for the fact that a "rogue" primary that keeps writing to the shard directory but is otherwise disconnected from the master will simply write a `index-${uuid}` blob that master does not take into account and that just becomes stale data to be cleaned up on the next delete instead of wreaking havoc on subsequent concurrent snapshots.